### ISSUE TITLE
cannot construct a valid set refer to a train set with max_bin != 255

### ISSUE THREAD

**Issue by aslongaspossible** [microsoft/LightGBM#6159]
## Description

When I want to construct a valid set refer to a train set with max_bin=15, it raises "Dataset max_bin 15 != config 255". Seems that I can never create a valid set with max_bin != 255?

## Reproducible example

```python
val = lgb.Dataset(val_dataframe, val_label, lgb.Dataset('train.bin'))
val.save_binary('val.bin')
```
Where `val_dataframe` are features of valid set, `val_label` are labels, `'train_bin'` are saved train set of `lgb.Dataset` in binary with `max_bin=15`.

## Environment info

LightGBM version or commit hash: 3.3.5

Command(s) you used to install LightGBM

```shell
conda install -c conda-forge lightgbm
```

**Comment by jameslamb** [comment]
Thanks for using LightGBM.

The reproducible example you provided isn't reproducible. For example:

* it's missing import statements
* you haven't provided code that could exactly reproduce the objects `val_dataframe` and `val_label`
* you haven't provided the content of the file `train.bin` or described exactly how it was created
* the name `val_dataframe` implies that you're using a data frame in memory, but you haven't mentioned what library (`pandas`? `polars`? `dask`?) or the version of that library

You could help reduce the effort required to answer this question by addressing those concerns and providing a minimal, reproducible example. If you haven't done that before and are unsure where to start, see:

* https://stackoverflow.com/help/minimal-reproducible-example
* any of the issues in this project with Python examples using the datasets from `sklearn.datasets`, e.g. https://github.com/microsoft/LightGBM/issues/4883#issuecomment-1003622596

Can you please provide such details or explain why that's not possible?

**Comment by aslongaspossible** [comment]
> Thanks for using LightGBM.
> 
> The reproducible example you provided isn't reproducible. For example:
> 
> * it's missing import statements
> * you haven't provided code that could exactly reproduce the objects `val_dataframe` and `val_label`
> * you haven't provided the content of the file `train.bin` or described exactly how it was created
> * the name `val_dataframe` implies that you're using a data frame in memory, but you haven't mentioned what library (`pandas`? `polars`? `dask`?) or the version of that library
> 
> You could help reduce the effort required to answer this question by addressing those concerns and providing a minimal, reproducible example.

Sorry to have provided unreproducible pseudocode. Here is the reproducible one:

```python
import numpy as np
import lightgbm as lgb
lgb_random = lgb.Dataset(np.random.rand(10000, 100), np.random.rand(10000, 1), params={'max_bin':15})
lgb_random.save_binary('random.bin')
del lgb_random
val = lgb.Dataset(np.random.rand(1000, 100), np.random.rand(1000, 1), reference=lgb.Dataset('random.bin'))
val.save_binary('val.bin')
``` 

In other words, any random data can reproduce my problem. I wonder that why the `config` doesn't change according to reference dataset.

**Comment by shiyu1994** [comment]
@aslongaspossible Thanks for reporting this issue.

It seems that by default the `lgb.Dataset` considers `max_bin` as `255` for default value. And even when loading from binary dataset file, the `max_bin` isn't loaded from the file.

A quick fix would be add `params={'max_bin':15}` in the `reference=lgb.Dataset('random.bin')`, i.e. `reference=lgb.Dataset('random.bin', params={'max_bin':15})`.

Still, I agree that this is not convenient. Since the binary file should ideally contain all the information to reconstruct the preprocessed dataset.

Will look into how to fix this laster.

**Comment by jameslamb** [comment]
Thanks @aslongaspossible for providing a reproducible example. Given that, I see the issue and agree with @shiyu1994 's recommendation. To facilitate that in the future, until LightGBM provides more convenient behavior, consider storing Dataset parameters alongside wherever you store the `Dataset` binary files.

> will look into how to fix this

@shiyu1994 @aslongaspossible please see https://github.com/microsoft/LightGBM/issues/4904#issuecomment-1003805133 where I described this exact issue in detail. I think we have a path forward, but haven't as yet had anyone take up implementing it: https://github.com/microsoft/LightGBM/issues/4904#issuecomment-1328470169

**Comment by aslongaspossible** [comment]
> A quick fix would be add `params={'max_bin':15}` in the `reference=lgb.Dataset('random.bin')`, i.e. `reference=lgb.Dataset('random.bin', params={'max_bin':15})`.

This works. Thank you!
